explorer/interpreter/type_checker.cpp

```diff
@@ -1327,49 +1363,123 @@ class TypeChecker::ConstraintTypeBuilder {
+                                     << *rewrite.converted_replacement << "\n";
+      }
+      rewrite_queue.Push(&rewrite);
     }
 
-    // Rewrite `Self.X is Y` to `Replacement is Y` if we have a rewrite for
-    // `Self.X`.
-    // TODO: Properly apply rewrites throughout all the constraints. Check for
-    // cycles. This is just a very short-term hack.
-    for (auto& impl_constraint : impl_constraints_) {
-      bool performed_rewrite;
-      do {
-        performed_rewrite = false;
-        if (const auto* assoc =
-                dyn_cast<AssociatedConstant>(impl_constraint.type)) {
-          for (const auto& rewrite : rewrite_constraints_) {
-            if (ValueEqual(assoc, rewrite.constant, std::nullopt)) {
-              impl_constraint.type = rewrite.converted_replacement;
-              performed_rewrite = true;
-            }
-          }
+    // Apply rewrites to each other and find a fixed point.
+    int rewrite_iterations = 0;
+    while (auto rewrite_entry = rewrite_queue.Pop()) {
+      auto* rewrite = *rewrite_entry;
+
+      // Track that we are rewriting this rewrite.
+      RewriteInfo info = {.rewrite = rewrite};
+      currently_rewriting_ = &info;
+      auto clear_currently_rewriting =
+          llvm::make_scope_exit([&] { currently_rewriting_ = std::nullopt; });
+
+      // Rebuild the rewrite and see if it changed.
+      Nonnull<const Value*> rebuilt = rebuild(rewrite->converted_replacement);
+      if (info.rewrite_used) {
+        // This would result in an infinite loop.
+        return ProgramError(source_loc)
+               << "rewrite of " << *rewrite->constant
+               << " applies within its own resolved expansion of " << *rebuilt;
+      }
+      if (!ValueEqual(rebuilt, rewrite->converted_replacement, std::nullopt)) {
+        if (type_checker.trace_stream_) {
+          **type_checker.trace_stream_ << "rewrote rewrite of "
+                                       << *rewrite->constant << " to "
+                                       << *rebuilt << "\n";
         }
-      } while (performed_rewrite);
+        rewrite->converted_replacement = rebuilt;
+        // Now we've rewritten this rewrite, we might find more rewrites apply
+        // to the portion we rewrote.
+        rewrite_queue.Push(rewrite);
+      } else {
+        if (type_checker.trace_stream_) {
+          **type_checker.trace_stream_ << "rewrite of " << *rewrite->constant
+                                       << " converged to " << *rebuilt << "\n";
+        }
+      }
+      ++rewrite_iterations;
+      // This iteration limit exists only to prevent large fuzzer-generated
+      // examples from leading to long compile times. If the limit is hit in a
+      // real example, it should be increased.
+      if (rewrite_iterations > 1000) {
+        return ProgramError(source_loc)
+               << "reached iteration limit resolving rewrite constraints";
+      }
```

Comment:
FWIW, given rewrite_iterations is pretty local logic, it might be nice to have it closer to the `int rewrite_iterations = 0;`. It would also not run the replacement logic in this scenario (which is pretty minor, given the issue it's checking for, but it still seems a little odd to me to do the work then discard it for an error that was expected at the start of the loop).